flink-ml-lib/src/main/java/org/apache/flink/ml/classification/knn/KnnModel.java

```diff
@@ -0,0 +1,412 @@
+            return Tuple2.of(items, metrics);
+        }
+
+        /**
+         * @param input input node.
+         * @param topN top N.
+         * @param priorityQueue priority queue.
+         */
+        private void search(
+                Object input,
+                Integer topN,
+                PriorityQueue<Tuple2<Double, Object>> priorityQueue,
+                KnnModelData modelData) {
+            Tuple2<DenseVector, Double> sample = computeNorm((DenseVector) input);
+            Tuple2<Double, Object> head = null;
+            for (int i = 0; i < modelData.getLength(); i++) {
+                ArrayList<Tuple2<Double, Object>> values = computeDistance(sample, i);
+                for (Tuple2<Double, Object> currentValue : values) {
+                    head = updateQueue(priorityQueue, topN, currentValue, head);
+                }
+            }
+        }
+
+        /**
+         * Updates queue.
+         *
+         * @param map queue.
+         * @param topN top N.
+         * @param newValue new value.
+         * @param head head value.
+         * @param <T> id type.
+         * @return head value.
+         */
+        private <T> Tuple2<Double, T> updateQueue(
+                PriorityQueue<Tuple2<Double, T>> map,
+                int topN,
+                Tuple2<Double, T> newValue,
+                Tuple2<Double, T> head) {
+            if (map.size() < topN) {
+                map.add(Tuple2.of(newValue.f0, newValue.f1));
+                head = map.peek();
+            } else {
+                if (map.comparator().compare(head, newValue) < 0) {
+                    Tuple2<Double, T> peek = map.poll();
+                    assert peek != null;
+                    peek.f0 = newValue.f0;
+                    peek.f1 = newValue.f1;
+                    map.add(peek);
+                    head = map.peek();
+                }
+            }
+            return head;
+        }
+
+        /**
+         * Computes distance between sample and dictionary vectors.
+         *
+         * @param input sample with l2 norm.
+         * @param index dictionary vectors index.
+         * @return distances.
+         */
+        private ArrayList<Tuple2<Double, Object>> computeDistance(
+                Tuple2<DenseVector, Double> input, Integer index) {
+            Tuple3<DenseMatrix, DenseVector, String[]> data = modelData.getDictData().get(index);
+
+            DenseMatrix res = calc(input, data);
+            ArrayList<Tuple2<Double, Object>> list = new ArrayList<>(0);
+            String[] curLabels = data.f2;
+            for (int i = 0; i < Objects.requireNonNull(curLabels).length; i++) {
+                Tuple2<Double, Object> tuple = Tuple2.of(res.values[i], curLabels[i]);
+                list.add(tuple);
+            }
+            return list;
+        }
+
+        /** The blas used to accelerating speed. */
+        private static final dev.ludovic.netlib.blas.F2jBLAS NATIVE_BLAS =
```

Comment:
We will need to put BLAS-related options in a shared infra class file such as `BLAS.java`. Both the NaiveBayes and LogisticRegression PR does this.

Maybe you could rebase your PR on e.g. the LogisticRegression PR and add the `dgemv` method as appropriate?

Reply:
done